File: app/javascript/channels/room_channel.js
import consumer from "./consumer"

document.addEventListener("turbolinks:load", function(){
  const chatChannel = consumer.subscriptions.create("RoomChannel", {

    connected() {
      // Called when the subscription is ready for use on the server
    },
  
    disconnected() {
      // Called when the subscription has been terminated by the server
    },
  
    received: function(data) {
      var room = document.getElementById('room');
      let chat_window = document.getElementById('chat_window');
      if(room){
        let createHTML = `<div class="message_left">
                            <div class="other_users_message message">
                              <div>
                                <p>${data['message']}</p>
                                <p>${data['created_at']}</p>
                              </div>
                            </div>
                          </div>`;
        $('#chat_window').append(createHTML);
        scrollBottom(chat_window);
        createHTML = '';
      }
      //return alert(data['message']);
    },
  
    speak: function(data) {
      return this.perform('speak',{
        data
      })
    },

    cookies: function(){
      return this.perform('cookies')
    },

    loadUserData: function(){
      return this.perform('subscribed')
    }

  });

  //チャットウィンドウ
  let window = document.getElementById('chat_window');
  //トークボタン
  const talk_button = document.querySelector('.talk_button');

  //トークのクリック時、room_channel.rbのsubscribedを実行
  if(talk_button){
    talk_button.addEventListener('click', function(){
      chatChannel.loadUserData();
    });
  }


  //一番下までスクロール
  function scrollBottom(object){
    object.scroll(0,10000000)
  };




  //チャットルームのdata属性からユーザーのデータを取得
  function getRoomData(message){
    var current_user_id = $('#room').data('current_user')
    var other_user_id = $('#room').data('other_user')
    var room_id = $('#room').data('room_id')
    var data = { current_user_id: current_user_id, other_user_id: other_user_id, room_id: room_id, message: message }
    return data
  }

  chatChannel.cookies();

  
  //数字が一桁であれば０で埋める
  var toDoubleDigits = function(num) {
    num += "";
    if (num.length === 1) {
      num = "0" + num;
    }
   return num;
  } 
  

  //現在時刻を取得
  
  function get_current_time(){
    
    var date = new Date();
    var year = date.getFullYear();
    var month = toDoubleDigits(date.getMonth() + 1);
    var day = toDoubleDigits(date.getDate());
    var hour = toDoubleDigits(date.getHours());
    var minute = toDoubleDigits(date.getMinutes());
    return `${year}/${month}/${day} ${hour}:${minute}`;
  };
  

  // //メッセージの送信・自分のルームにメッセージを非同期で追加
  
  let talk_room = document.querySelector('.talk_room');
  if(talk_room != null){
    var send_message_form = document.getElementById('send_message_form');
    var send_message_btn = document.getElementById('send_message_btn');
    send_message_btn.addEventListener('click', function(){
      var content = send_message_form.value;
      if(1 <= content.length　&& content.length < 140){
        var createHTML = `<div class="message_right">
                          <div class="users_message message">
                            <div>
                                <p>${content}</p>
                                <p>${get_current_time()}</p>
                            </div>
                          </div>
                        </div>`
      $('#chat_window').append(createHTML);
      let chat_window = document.getElementById('chat_window');
      scrollBottom(chat_window);
      let room_data = getRoomData(content);
      chatChannel.speak(room_data);
      send_message_form.value = '';
      createHTML = '';
      }else if(content.length >= 140){
        send_message_form.value = '';
      }
    })
  }
})

File: app/channels/room_channel.rb
class RoomChannel < ApplicationCable::Channel

  def subscribed
    @user = User.find_by(id: cookies.signed[:user_id])
    if @user
      #ユーザのルームを全て取得
      @rooms = @user.rooms.where(user_id: @user.id)
      #ルームが存在していればルームIDのチャネルを生成
      if @rooms
        @room_array = @rooms.map{ |room| "room_#{room.id}" }
        @room_array.each do |channel_name|
          stream_from "#{channel_name}"
        end
      end
    else
    end
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def speak(data)
    @room_data = data['data']

    @room_id = @room_data['room_id']
    @user_id = @room_data['current_user_id']
    @other_user_id = @room_data['other_user_id']
    @content = @room_data['message']

    #自分
    @user = User.find(@user_id)
    #相手
    @other_user = User.find(@other_user_id)
    #メッセージを自分のルームに保存
    @message = @user.messages.new(room_id: @room_id ,user_id: @user.id, other_user_id: @other_user.id,
                                          content: @content)
    if @message.save
      #相手に自分とのメッセージルームがあるか確認
      @other_user_room = Room.find_by(user_id: @other_user.id, other_user_id: @user.id)
      if @other_user_room #ルームが存在していればメッセージを保存
        @other_message = Message.new(room_id: @other_user_room.id, user_id: @user.id, other_user_id: @other_user.id, content: @content)
        @other_message.save
        send_message_with_action_cable(@other_message)
      else #無ければ新しくルームを作成してメッセージを保存
        @create_other_user_room = Room.new(user_id: @other_user.id, other_user_id: @user.id)
        @create_other_user_room.save
        @other_message = Message.new(room_id: @create_other_user_room.id, user_id: @user.id , 
                                     other_user_id: @other_user.id , content: @content)
        @other_message.save
        send_message_with_action_cable(@other_message)
      end
      
    else
    end
    
  end

  
  def cookies()
    connection.send(:cookies)
  end
    private

    def send_message_with_action_cable(message)
      ActionCable.server.broadcast "room_#{message.room_id}", message: message.content, created_at: message.created_at
    end

end